src/dev/encoder.cpp

```diff
@@ -0,0 +1,152 @@
+#include <EVT/dev/encoder.hpp>
+
+namespace EVT::core::DEV {
+
+Encoder::Encoder(IO::GPIO& a, IO::GPIO& b, int64_t range, int64_t initialPosition) : a(a), b(b), range(range), position(initialPosition) {
+    //making sure range is positive
+    if (range < 0) {
+        range *= -1;
+    }
+    setPosition(position);
+    //setting other instance variables
+    currentDirection = Static;
+    noChangeCounter = 0;
+    currentRelPos = readPinValues();
+}
+
+int64_t Encoder::getPosition() {
+    return position;
+}
+
+int8_t Encoder::update() {
+    //Reads current pin values
+    int8_t newPos = readPinValues();
+    //find the change in values;
+    int8_t change = calculateChange(newPos);
+    //sets the position to its new value
+    setPosition(position + change);
+    return change;
+}
+
+int8_t Encoder::calculateChange(int8_t newRelPos) {
+    //Gets the change between the positions and then sets currentRelPos to the newRelPos
+    int8_t change = newRelPos - currentRelPos;
```

Comment:
I don't think just subtracting makes sense here because it doesn't have enough information. If the current position is 2, the current direction is right, and it is turned 3 to 1, you're going to get -1 here, which will turn into 1 in the if statements at the bottom of the function

Reply:
In any situation in which the encoder is turned 3 ticks between updates, it's impossible to tell between 3 in one direction or -1 in the other. It makes more sense to assume that it was 1 in the other direction, since that is the more likely scenario.